src/ILCompiler.DependencyAnalysisFramework/src/DependencyNodeCore.cs

```diff
@@ -67,9 +67,35 @@ public struct CombinedDependencyListEntry
             }
 
             // Used by HashSet, so must have good Equals/GetHashCode
-            public DependencyNodeCore<DependencyContextType> Node;
-            public DependencyNodeCore<DependencyContextType> OtherReasonNode;
-            public string Reason;
+            public readonly DependencyNodeCore<DependencyContextType> Node;
+            public readonly DependencyNodeCore<DependencyContextType> OtherReasonNode;
+            public readonly string Reason;
+
+            public override bool Equals(object obj)
+            {
+                return obj is CombinedDependencyListEntry && Equals((CombinedDependencyListEntry)obj);
+            }
+
+            public override int GetHashCode()
+            {
+                int hash = 23;
+                hash = hash * 31 + Node.GetHashCode();
+
+                if (OtherReasonNode != null)
+                    hash = hash * 31 + OtherReasonNode.GetHashCode();
+
+                if (Reason != null)
+                    hash = hash * 31 + Reason.GetHashCode();
+
+                return hash;
+            }
+
+            public bool Equals(CombinedDependencyListEntry other)
+            {
+                return Object.Equals(Node, other.Node)
```

Comment:
The DependencyNode's don't quite work unless the nodes are single instanced and reference comparable. I think you could just use == or Object.ReferenceEquals here. (We should probably also make DependencyNode override Equals to do the same and make it sealed.)